File: src/Circle.AI.Personality/PersonaPromptBuilder.cs
// PersonaPromptBuilder.cs
//
// Renders a Persona into a compact natural-language system-prompt hint.
// Defensive against prompt-injection: every user-controlled string is
// emitted as a JSON string literal so any embedded quotes or newlines are
// escaped before reaching the model.

using System.Text;
using System.Text.Encodings.Web;
using System.Text.Json;

namespace Circle.AI.Personality;

/// <summary>
/// Builds the natural-language system-prompt block describing a
/// <see cref="Persona"/>. Returns an empty string when the persona is in
/// its default/unedited state so the prompt is not bloated with no-op
/// instructions.
/// </summary>
public static class PersonaPromptBuilder
{
    private static readonly JsonSerializerOptions s_stringQuote = new()
    {
        // Hardened escaping — keep control characters and quotes inside the
        // JSON-encoded literal so a malicious taboo entry cannot break out
        // of the [Persona] block.
        Encoder = JavaScriptEncoder.UnsafeRelaxedJsonEscaping,
    };

    /// <summary>
    /// Renders <paramref name="persona"/> into a compact system-prompt hint.
    /// </summary>
    /// <param name="persona">The persona to render. Required.</param>
    /// <returns>
    /// A persona-block string, or an empty string when the persona is
    /// effectively default (display name only).
    /// </returns>
    public static string BuildSystemHint(Persona persona)
    {
        ArgumentNullException.ThrowIfNull(persona);

        if (IsEffectivelyDefault(persona)) return string.Empty;

        var sb = new StringBuilder();
        sb.Append("[Persona]");

        sb.Append("\nYou are speaking with ");
        sb.Append(Quote(persona.DisplayName));
        sb.Append('.');

        if (!string.IsNullOrWhiteSpace(persona.Pronouns))
        {
            sb.Append(" They identify as ");
            sb.Append(Quote(persona.Pronouns));
            sb.Append('.');
        }

        sb.Append("\nThey prefer responses in ");
        sb.Append(Quote(persona.PreferredLocale));
        sb.Append(", tone between ");
        sb.Append(Quote(persona.Formality.Floor));
        sb.Append(" and ");
        sb.Append(Quote(persona.Formality.Ceiling));
        sb.Append('.');

        if (persona.IdentityTags.Count > 0)
        {
            sb.Append("\nIdentity tags: ");
            sb.Append(QuoteList(persona.IdentityTags));
            sb.Append('.');
        }

        if (persona.Values.Count > 0)
        {
            sb.Append("\nTheir declared values: ");
            sb.Append(QuoteList(persona.Values));
            sb.Append('.');
        }

        if (persona.Taboos.Count > 0)
        {
            sb.Append("\nAvoid: ");
            sb.Append(QuoteList(persona.Taboos));
            sb.Append('.');
        }

        if (!string.IsNullOrWhiteSpace(persona.VoicePreference))
        {
            sb.Append("\nPreferred voice tag: ");
            sb.Append(Quote(persona.VoicePreference));
            sb.Append('.');
        }

        if (persona.Privacy == PrivacyLevel.Strict)
        {
            sb.Append("\nPrivacy: strict — minimize stored signals, do not surface personal context proactively, and never share personal context across surfaces without explicit prompt.");
        }
        else if (persona.Privacy == PrivacyLevel.Open)
        {
            sb.Append("\nPrivacy: open — the user has authorised broader retention and proactive surfacing.");
        }

        return sb.ToString();
    }

    // ------------------------------------------------------------------
    // Helpers
    // ------------------------------------------------------------------

    /// <summary>
    /// True when the persona contains no information beyond the
    /// <see cref="Persona.Create(string, string)"/> defaults.
    /// </summary>
    private static bool IsEffectivelyDefault(Persona p) =>
        string.IsNullOrWhiteSpace(p.Pronouns)
        && p.IdentityTags.Count == 0
        && p.Values.Count == 0
        && p.Taboos.Count == 0
        && string.IsNullOrWhiteSpace(p.VoicePreference)
        && p.Privacy == PrivacyLevel.Balanced
        && string.Equals(p.Formality.Floor, "casual", StringComparison.Ordinal)
        && string.Equals(p.Formality.Ceiling, "formal", StringComparison.Ordinal);

    /// <summary>
    /// JSON-encodes <paramref name="value"/> into a quoted literal. This is
    /// the prompt-injection defence: any embedded quote, newline, or
    /// directive ("ignore previous instructions") is rendered as inert text
    /// inside a quoted string.
    /// </summary>
    private static string Quote(string value) =>
        JsonSerializer.Serialize(value, s_stringQuote);

    private static string QuoteList(IReadOnlyList<string> items)
    {
        if (items.Count == 0) return string.Empty;
        var parts = new string[items.Count];
        for (int i = 0; i < items.Count; i++) parts[i] = Quote(items[i]);
        return string.Join(", ", parts);
    }
}

File: python/src/circle_ai/personality/persona_prompt_builder.py
# persona_prompt_builder.py
#
# Port of CircleAI.Personality PersonaPromptBuilder.cs (C# — the EXACT spec).
#
# Renders a Persona into a compact natural-language system-prompt hint. Returns
# an empty string when the persona is in its default/unedited state so the prompt
# is not bloated with no-op instructions.
#
# Prompt-injection defence (faithful port): every user-controlled string is
# emitted as a JSON string literal so any embedded quote / newline / directive is
# inert text inside a quoted string. The C# uses
# JsonSerializer.Serialize(value, {Encoder = UnsafeRelaxedJsonEscaping}); Python's
# json.dumps(value, ensure_ascii=False) matches that relaxed escaping (quotes,
# backslash and control chars escaped; <, >, &, + left as-is).

from __future__ import annotations

import json
from typing import Sequence

from .persona import Persona, PrivacyLevel


def _quote(value: str) -> str:
    """JSON-encode ``value`` into a quoted literal (relaxed escaping)."""
    return json.dumps(value, ensure_ascii=False)


def _quote_list(items: Sequence[str]) -> str:
    if len(items) == 0:
        return ""
    return ", ".join(_quote(i) for i in items)


def _is_effectively_default(p: Persona) -> bool:
    """True when the persona contains no information beyond the
    :meth:`Persona.create` defaults."""
    pronouns = p.pronouns
    voice = p.voice_preference
    return (
        (pronouns is None or str(pronouns).strip() == "")
        and len(p.identity_tags) == 0
        and len(p.values) == 0
        and len(p.taboos) == 0
        and (voice is None or str(voice).strip() == "")
        and p.privacy == PrivacyLevel.BALANCED
        and p.formality.floor == "casual"
        and p.formality.ceiling == "formal"
    )


def build_system_hint(persona: Persona) -> str:
    """Render ``persona`` into a compact system-prompt hint, or an empty string
    when the persona is effectively default. Mirrors
    ``PersonaPromptBuilder.BuildSystemHint``."""
    if persona is None:
        raise ValueError("persona")

    if _is_effectively_default(persona):
        return ""

    parts = ["[Persona]"]
    parts.append("\nYou are speaking with ")
    parts.append(_quote(persona.display_name))
    parts.append(".")

    pronouns = persona.pronouns
    if pronouns is not None and str(pronouns).strip() != "":
        parts.append(" They identify as ")
        parts.append(_quote(str(pronouns)))
        parts.append(".")

    parts.append("\nThey prefer responses in ")
    parts.append(_quote(persona.preferred_locale))
    parts.append(", tone between ")
    parts.append(_quote(persona.formality.floor))
    parts.append(" and ")
    parts.append(_quote(persona.formality.ceiling))
    parts.append(".")

    if len(persona.identity_tags) > 0:
        parts.append("\nIdentity tags: ")
        parts.append(_quote_list(persona.identity_tags))
        parts.append(".")

    if len(persona.values) > 0:
        parts.append("\nTheir declared values: ")
        parts.append(_quote_list(persona.values))
        parts.append(".")

    if len(persona.taboos) > 0:
        parts.append("\nAvoid: ")
        parts.append(_quote_list(persona.taboos))
        parts.append(".")

    voice = persona.voice_preference
    if voice is not None and str(voice).strip() != "":
        parts.append("\nPreferred voice tag: ")
        parts.append(_quote(str(voice)))
        parts.append(".")

    if persona.privacy == PrivacyLevel.STRICT:
        parts.append(
            "\nPrivacy: strict — minimize stored signals, do not surface personal "
            "context proactively, and never share personal context across surfaces "
            "without explicit prompt."
        )
    elif persona.privacy == PrivacyLevel.OPEN:
        parts.append(
            "\nPrivacy: open — the user has authorised broader retention and "
            "proactive surfacing."
        )

    return "".join(parts)


class PersonaPromptBuilder:
    """Static-style holder mirroring the C# ``PersonaPromptBuilder`` class."""

    build_system_hint = staticmethod(build_system_hint)
